### PR TITLE
bugfix for sherpa compiled with openmpi support

### PR DESCRIPTION
**Pull request by pmillet**

### CODE REVIEW

**Comment by perrozzi** [comment]
@pmillet please apply the suggested patch https://cmssdt.cern.ch/SDT/code-checks/PR-20401/525/git-diff.patch and commit/push again

**Comment by perrozzi** [comment]
please test

**Comment by pmillet** [comment]
If I checkout CMSSW_9_4_X_2017-09-06-2300 and apply the changes to the SherpaInterface, scram b is working fine.
From the error message this maybe is related to https://github.com/cms-sw/cmssw/commit/d0cc21f6f7f18d179e626f33041dde5f512e7c52 ?

**Comment by perrozzi** [comment]
please test

**Comment by perrozzi** [comment]
@pmillet just fo me to understand: is this fixing the event generation when cmssw is compiled with mpi, so that both sherpack and event generation will be ok?

**Comment by pmillet** [comment]
@perrozzi It is supposed to fix exactly that and worked fine in local testing with cmsRun in CMSSW_93DEVELX

**Comment by smuzaffar** [comment]
please test workflow 534.0
534.0 should still fail (due to gridpacks GCC ABI incompatibility issue) but it should not fail with error message
```
*** The MPI_Comm_rank() function was called before MPI_INIT was invoked.
*** This is disallowed by the MPI standard.
*** Your MPI job will now abort.
[cmsbuild06.cern.ch:9175] Local abort before MPI_INIT completed completed successfully, but am not able to aggregate error messages, and not able to guarantee that all other processes were killed!
```

**Comment by pmillet** [comment]
So if I read the log correctly we are back to the gcc incompatibility error as expected. I will update the relval samples for gcc630.

``` shell
Library_Loader::LoadLibrary(): /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/runTheMatrix-results/534.0_sherpa_ZtoEE_0j_OpenLoops_13TeV_MASTER+sherpa_ZtoEE_0j_OpenLoops_13TeV_MASTER+HARVESTGEN/Process/Amegic/lib/libProc_fsrchannels2.so: undefined symbol: _ZN6PHASIC14Single_Channel4NameEv
Library_Loader::LoadLibrary(): Failed to load library 'libProc_fsrchannels2.so'.
```

**Comment by pmillet** [comment]
anything preventing this from being included? its needed for Sherpa to work with https://github.com/cms-sw/cmsdist/pull/3405 (merged)

**Comment by perrozzi** [comment]
+1

**Comment by perrozzi** [comment]
I was actually waiting for https://github.com/cms-sw/cmssw/pull/20515 to be merged in order to re-run the tests. From the pure GEN point of view this is good to go, approved.

File: GeneratorInterface/SherpaInterface/src/SherpaHadronizer.cc [hunk 1]
```diff
@@ -66,7 +69,7 @@ class SherpaHadronizer : public gen::BaseHadronizer {
   edm::ParameterSet  SherpaParameterSet;
   unsigned int maxEventsToPrint;
   std::vector<std::string> arguments;
-  SHERPA::Sherpa Generator;
+  SHERPA::Sherpa *Generator = new SHERPA::Sherpa();
```
Comment (davidlange6) [inline GeneratorInterface/SherpaInterface/src/SherpaHadronizer.cc]: hi @pmillet - please make this a smart pointer.

File: GeneratorInterface/SherpaInterface/src/SherpaHadronizer.cc [hunk 2]
```diff
@@ -186,17 +188,21 @@ SherpaHadronizer::SherpaHadronizer(const edm::ParameterSet &params) :
...
+  arguments.push_back(shRng);
   isInitialized=false;
  //initialization of Sherpa moved to initializeForInternalPartons
 }
 
 SherpaHadronizer::~SherpaHadronizer()
 {
+  Generator->~Sherpa();
+  #ifdef USING__MPI
+    MPI::Finalize();
```
Comment (davidlange6) [inline GeneratorInterface/SherpaInterface/src/SherpaHadronizer.cc]: so this is a sign that you can not use the default CMSSW builds.. have you considered run time switches to provide this same functionality?

Comment (pmillet) [reply GeneratorInterface/SherpaInterface/src/SherpaHadronizer.cc]: sherpa is either compiled with or without mpi support, so runtime switches are not an option
if it is compiled with mpi support it needs the mpi init and finalize calls even if it is not parallelized using mpi (this is the case here)

**Comment by davidlange6** [comment]
merge

**Pull request merged**